File: python/lsst/meas/algorithms/coaddBoundedField.cc
/*
 * LSST Data Management System
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 * See the COPYRIGHT file
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <https://www.lsstcorp.org/LegalNotices/>.
 */
#include "pybind11/pybind11.h"
#include "lsst/cpputils/python.h"
#include "pybind11/stl.h"

#include "lsst/geom/Box.h"
#include "lsst/afw/table/io/python.h"
#include "lsst/meas/algorithms/CoaddBoundedField.h"

namespace py = pybind11;
using namespace pybind11::literals;

namespace lsst {
namespace meas {
namespace algorithms {
namespace {

void declareCoaddBoundedField(lsst::cpputils::python::WrapperCollection &wrappers) {
    using PyElement = py::class_<CoaddBoundedFieldElement>;
    static auto clsElement = wrappers.wrapType(PyElement(wrappers.module, "CoaddBoundedFieldElement"), [](auto &mod, auto &cls) {
        cls.def(py::init([](std::shared_ptr<afw::math::BoundedField> field,
                            std::shared_ptr<afw::geom::SkyWcs const> wcs, py::object polygon, double weight) {
                    if (polygon.is(py::none())) {
                        return new CoaddBoundedFieldElement(field, wcs, nullptr, weight);
                    } else {
                        auto pgon = py::cast<std::shared_ptr<afw::geom::polygon::Polygon const>>(polygon);
                        return new CoaddBoundedFieldElement(field, wcs, pgon, weight);
                    }
                }),
                "field"_a, "wcs"_a, "validPolygon"_a, "weight"_a = 1.0);
        cls.def_readwrite("field", &CoaddBoundedFieldElement::field);
        cls.def_readwrite("wcs", &CoaddBoundedFieldElement::wcs);
        cls.def_readwrite("validPolygon", &CoaddBoundedFieldElement::validPolygon);
        cls.def_readwrite("weight", &CoaddBoundedFieldElement::weight);

        cls.def("__eq__", &CoaddBoundedFieldElement::operator==, py::is_operator());
        cls.def("__ne__", &CoaddBoundedFieldElement::operator!=, py::is_operator());
    });
    using PyClass = py::class_<CoaddBoundedField, afw::math::BoundedField>;
    auto clsField = wrappers.wrapType(PyClass(wrappers.module, "CoaddBoundedField"), [](auto &mod, auto &cls) {

        cls.attr("Element") = clsElement;

        /* Constructors */
        cls.def(py::init<geom::Box2I const &, std::shared_ptr<afw::geom::SkyWcs const>,
                                         typename CoaddBoundedField::ElementVector const &>(),
                                 "bbox"_a, "coaddWcs"_a, "elements"_a);
        cls.def(py::init<geom::Box2I const &, std::shared_ptr<afw::geom::SkyWcs const>,
                                         typename CoaddBoundedField::ElementVector const &, double>(),
                                 "bbox"_a, "coaddWcs"_a, "elements"_a, "default"_a);

        /* Operators */
        cls.def("__eq__", &CoaddBoundedField::operator==, py::is_operator());
        cls.def("__ne__", &CoaddBoundedField::operator!=, py::is_operator());
        cls.def("__imul__", &CoaddBoundedField::operator*);

        /* Members */
        cls.def("_evaluate", &CoaddBoundedField::evaluate);
        cls.def("getCoaddWcs", &CoaddBoundedField::getCoaddWcs);
        cls.def("getDefault", &CoaddBoundedField::getDefault);
        cls.def("getElements", &CoaddBoundedField::getElements);
        cls.def("getThrowOnMissing", &CoaddBoundedField::getThrowOnMissing);
    });
    afw::table::io::python::addPersistableMethods<CoaddBoundedField>(clsField);
}

}  // namespace
void wrapCoaddBoundedField(lsst::cpputils::python::WrapperCollection &wrappers) {
    declareCoaddBoundedField(wrappers);
}
}  // namespace algorithms
}  // namespace meas
}  // namespace lsst

File: python/lsst/meas/algorithms/coaddBoundedField.py
# This file is part of meas_algorithms.
#
# Developed for the LSST Data Management System.
# This product includes software developed by the LSST Project
# (https://www.lsst.org).
# See the COPYRIGHT file at the top-level directory of this distribution
# for details of code ownership.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import numpy as np

from lsst.geom import Point2D
from lsst.utils import continueClass

from ._algorithmsLib import CoaddBoundedField

__all__ = ["CoaddBoundedField"]


@continueClass
class CoaddBoundedField:  # noqa: F811
    def evaluate(self, x, y=None):
        """Evaluate the CoaddBoundedField.

        This accepts either a Point2D or an array of x and y
        positions.

        When arrays are passed, this uses a vectorized version
        of CoaddBoundedField::evaluate(). If the coadd bounded
        field has throwOnMissing then this will return NaN
        for missing values; otherwise it will return 0.0.

        Parameters
        ----------
        x : `lsst.geom.Point2D` or `np.ndarray`
            Array of x values.
        y : `np.ndarray`, optional
            Array of y values.

        Returns
        -------
        values : `float` or `np.ndarray`
            Evaluated value or array of values.
        """
        if isinstance(x, Point2D):
            return self._evaluate(x)

        _x = np.atleast_1d(x).ravel()
        _y = np.atleast_1d(y).ravel()

        if len(_x) != len(_y):
            raise ValueError("x and y arrays must be the same length.")

        ra, dec = self.getCoaddWcs().pixelToSkyArray(_x, _y)

        sums = np.zeros(len(_x))
        wts = np.zeros_like(sums)

        for elt in self.getElements():
            ix, iy = elt.wcs.skyToPixelArray(ra, dec)
            in_box = elt.field.getBBox().contains(ix, iy)
            if elt.validPolygon:
                in_box &= elt.validPolygon.contains(ix, iy)
            sums[in_box] += elt.weight*elt.field.evaluate(ix[in_box], iy[in_box])
            wts[in_box] += elt.weight

        good = (wts > 0)
        values = np.zeros(len(_x))
        values[good] = sums[good]/wts[good]

        if self.getThrowOnMissing():
            values[~good] = np.nan

        return values
